### ISSUE TITLE
Cannot parse specific .DAT files

### ISSUE THREAD

**Issue by eouree** [Deep-Symmetry/crate-digger#10]
Hello, hope you are all well.
First of all I want to say that you did great work.
I tried to parse around 60000 analysis files (.DAT and .EXT files originally created not from Rekordbox but from DJ mixer CDJ 2000 NXS2) using C# code generated from the .ksy file. I have managed to parse most of the files successfully but for around 2500 .DAT files I got the following Exception Unhandled: System.IO.EndOfStreamException: 'requested 400 bytes, but got only 0 bytes' in “public byte[] ReadBytes(ulong count)” method.
The strange thing is that when am using Web IDE to see the structure and visually explore the contents of these files, I get no error and I can see them normally. Also am using the latest version of your code.
Can you please help me on this issue? I can provide you more samples of these .DAT files if you want.
Thanks

[DAT FILES.zip](https://github.com/Deep-Symmetry/crate-digger/files/4319162/DAT.FILES.zip)

**Comment by brunchboy** [comment]
Hello, thanks for the sample. I’m traveling due to a family medical emergency at the moment, so I can’t predict when I might have time to look at this, though. I also don’t use or have access to C#, so if the issue is specific to that, I won’t be able to help. Java is probably similar, though, so there is some hope. If you do figure it out in the mean time, a pull request would be welcome.

**Comment by eouree** [comment]
Hope all will go well with your emergency men

**Comment by b020** [comment]
Hello. I believe I am having the same issue but with java. I am trying to read .DAT files that was generated by Rekordbox, but it keeps throwing the following exception for every file I try to read PathTag:

```
java.nio.BufferUnderflowException
	at java.base/java.nio.DirectByteBuffer.get(DirectByteBuffer.java:299)
	at java.base/java.nio.ByteBuffer.get(ByteBuffer.java:804)
	at io.kaitai.struct.ByteBufferKaitaiStream.readBytes(ByteBufferKaitaiStream.java:274)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$PathTag._read(RekordboxAnlz.java:245)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$PathTag.<init>(RekordboxAnlz.java:240)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$PathTag.<init>(RekordboxAnlz.java:229)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$PathTag.fromFile(RekordboxAnlz.java:225)
```

Here is my code for the error above:
`RekordboxAnlz.PathTag pathTag = RekordboxAnlz.PathTag.fromFile(file.getPath());`

[00021195.zip](https://github.com/Deep-Symmetry/crate-digger/files/4327655/00021195.zip)

Actualy, I tried others Tags but it did not return any data correctly, although Web IDE gives me correct values also.

I am using Rekordbox 5.8.4.

Long story short, I am hooking up FreestylerDMX with Rekordbox. I managed to hack the pioneer software and gather some data on the fly from the software memory so I can automate some control of FreestylerDMX action. As soon as I get it working, I will publish and update you.

By the way, I hope everything is well about your emergency. =)

**Comment by brunchboy** [comment]
Thank you, things seem to be going well and I hope to be home soon.

It’s not surprising (and indeed reassuring) to hear that the same issue is happening in Java. The fix needs to happen in the `.ksy` file (the Kaitai Struct Compiler specification for the file format). If anyone can study that and the `.DAT` files and come up with a fix before I do, a pull request would be welcome.

**Comment by b020** [comment]
Hi. I guess that I was using it incorrectly.
Got it to read some files doing the following:

```
RekordboxAnlz anlz = new RekordboxAnlz(new ByteBufferKaitaiStream(file.getPath())); for(RekordboxAnlz.TaggedSection taggedSection : anlz.sections()){
     if(taggedSection.fourcc() == RekordboxAnlz.SectionTags.PATH.id()){
            System.out.println(((RekordboxAnlz.PathTag)taggedSection.body()).path());
     }
}
```

But now I got the following exception after a few reads:

```
Exception in thread "main" java.lang.NullPointerException
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$SongStructureEntry._read(RekordboxAnlz.java:769)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$SongStructureEntry.<init>(RekordboxAnlz.java:764)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$SongStructureTag._read(RekordboxAnlz.java:529)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$SongStructureTag.<init>(RekordboxAnlz.java:518)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$TaggedSection._read(RekordboxAnlz.java:1123)
	at org.deepsymmetry.cratedigger.pdb.RekordboxAnlz$TaggedSection.<init>(RekordboxAnlz.java:1083)
```

I will keep investigating!

**Comment by brunchboy** [comment]
Ah ha! Nice work!

**Comment by eouree** [comment]
The specific Exception Unhandled: System.IO.EndOfStreamException: 'requested 400 bytes, but got only 0 bytes' occurring when the program trying to parse the WavePreviewTags (1347900978 and 1347895638). For the rest tags seems its ok.

**Comment by brunchboy** [comment]
@b020 were you using a version that already had #6 merged in? Because that should handle phrase style 3.

**Comment by brunchboy** [comment]
@eouree I see the same exception using the Kaitai Web IDE to try to analyze your files, so the problem is definitely in the `.ksy` mapping file. Let’s take a look at those wave preview tags.

**Comment by brunchboy** [comment]
Studying the file you uploaded, @eouree, it is corrupt, and it makes sense that the parse is crashing. Those two wave preview tags (`PWAV` and `PWV2`) both report that the length of their header is `0x14` bytes, which is normal, but they also report that the length of the entire tag is `0x14` bytes, which means there is nothing beyond the header. But they then go on to claim that their are `0x190` (four hundred) bytes of waveform preview in the case of the `PWAV` tag, and `0x64` (one hundred) bytes of waveform preview in the case of the `PWV2` tag. But there are actually no preview bytes available in either tag, because the header is the only thing in the tag.

Can you explain in more detail how you are creating these files? Whatever is doing it is doing it incorrectly.

**Comment by brunchboy** [comment]
I have released a snapshot version, `0.1.4-SNAPSHOT` that avoids crashing in this situation. But it means that your code is going to have to handle the fact that some wave preview tags will return `nil` when you ask for their `data()`. I am updating Beat Link to handle this. Please let me know if this resolves your issue.

**Comment by eouree** [comment]
Hi there. These files were created on a CDJ2000. Structure of the files seems to be similar on those of Rekordbox with a few differences.

**Comment by brunchboy** [comment]
Can you please explain what you mean? I have never seen any evidence that CDJs can create files like this. I have a pair of CDJ 2000 nexus players. They can only use analysis files created by rekordbox, they cannot create their own, as far as I can tell. Can you point me at a manual or explanation of how one would even ask them to?

**Comment by brunchboy** [comment]
Also, did the new snapshot of crate-digger resolve your issue?

**Comment by eouree** [comment]
Yes the new snapshot of crate-digger is handling the issue. I have tested it also on C# code generated from the .ksy file. As for the evidence :)) check the path on the sample files i have uploaded before, its not like the path usually Rekordbox creates.

**Comment by brunchboy** [comment]
Thanks for the confirmation! I will mark this as resolved, but still wish you would explain how to get CDJs to create such files.

**Comment by eouree** [comment]
I will revert back to you once the quarantine measures finish so that i can have physical access to the mixers. Thank you

**Comment by brunchboy** [comment]
Thank you, stay safe!
